lib/HLSL/DxilReduceMSAAToSingleSample.cpp

```diff
@@ -0,0 +1,81 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// DxilReduceMSAAToSingleSample.cpp                                          //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Provides a pass to reduce all MSAA writes to single-sample writes         //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/HLSL/DxilGenerationPass.h"
+#include "dxc/HLSL/DxilOperations.h"
+#include "dxc/HLSL/DxilModule.h"
+#include "dxc/HLSL/DxilInstructions.h"
+
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/IR/Constants.h"
+
+using namespace llvm;
+using namespace hlsl;
+
+class DxilReduceMSAAToSingleSample : public ModulePass {
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilReduceMSAAToSingleSample() : ModulePass(ID) {}
+  const char *getPassName() const override { return "HLSL DXIL Reduce all MSAA writes to single-sample writes"; }
+  bool runOnModule(Module &M) override;
+};
+
+bool DxilReduceMSAAToSingleSample::runOnModule(Module &M)
+{
+  // This pass removes all instances of the discard instruction within the shader.
+  DxilModule &DM = M.GetOrCreateDxilModule();
+
+  LLVMContext & Ctx = M.getContext();
+  OP *HlslOP = DM.GetOP();
+
+  auto TextureLoadOverloads = std::vector<Type*>{ /*todo: what's fp16 type? */ Type::getFloatTy(Ctx), Type::getInt16Ty(Ctx), Type::getInt32Ty(Ctx) };
+
+  bool Modified = false;
+
+  for (const auto & Overload : TextureLoadOverloads) {
+
+    Function * TexLoadFunction = HlslOP->GetOpFunc(DXIL::OpCode::TextureLoad, Overload);
+    auto TexLoadFunctionUses = TexLoadFunction->uses();
+
+    for (auto FI = TexLoadFunctionUses.begin(); FI != TexLoadFunctionUses.end(); ) {
+      auto & FunctionUse = *FI++;
+      auto FunctionUser = FunctionUse.getUser();
+      auto instruction = cast<Instruction>(FunctionUser);
+      DxilInst_TextureLoad LoadInstruction(instruction);
+      auto TextureHandle = LoadInstruction.get_srv();
+      auto TextureHandleInst = cast<CallInst>(TextureHandle);
+      DxilInst_CreateHandle createHandle(TextureHandleInst);
+      // Dynamic rangeId is not supported 
+      if (isa<ConstantInt>(createHandle.get_rangeId())){
+        unsigned rangeId = cast<ConstantInt>(createHandle.get_rangeId())->getLimitedValue();
+        auto Resource = DM.GetSRV(rangeId);
```

Comment:
You're only checking for MS later, which means you still will call DM.GetSRV(rangeId) on a potentially invalid rangeId, and ask it if it's an MS texture.  In the test case I propose below, it will think it's an MS texture because it will look up SRV 0, instead of UAV 0.